Structure_Engine/Compute/ReinforcementDensity.cs

```diff
@@ -0,0 +1,114 @@
+﻿/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2021, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using BH.oM.Base; 
+using BH.oM.Structure.Elements;
+using BH.oM.Structure.Fragments;
+using BH.oM.Structure.MaterialFragments;
+using BH.oM.Structure.Results;
+using BH.oM.Reflection.Attributes;
+using BH.Engine.Base;
+using BH.Engine.Spatial;
+
+
+namespace BH.Engine.Structure
+{
+    public static partial class Compute
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+        [Description("Calculates the ReinforcementDensity of a BarRequiredArea from a given Bar and Material.")]
+        [Input("barRequiredArea", "The BarRequiredArea containing the reinforced areas and ids for the Bar and Material.")]
+        [Input("bars", "The Bars to search for the relevant Bar associated with the BarRequiredArea.")]
+        [Input("materials", "The Materials to search for the relevant Material associated with the BarRequiredArea.")]
+        [Output("reinforcementDensity", "The ReinforcementDensity calculated using the inputs provided.")]
+        public static ReinforcementDensity ReinforcementDensity(BarRequiredArea barRequiredArea, List<Bar> bars, List<IMaterialFragment> materials)
```

Comment:
As a first quick comment, would make this list take  list of `BarRequiredArea` as input and output a list of `ReinforcementDensity` with the same length.

This is mainly for efficiency reasons, as the current way the method is set up, the dictionaries etc have to be generated once for each `BarRequiredArea`.